### PR TITLE
Refactored KeepInterner to be more clear about notifications / side effects when keeping

### PR DESCRIPTION
**Pull request by andrewconner**
Tests compile, running now. @leogrim @ryanpbrewster

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala [hunk 1]
```diff
@@ -314,7 +314,8 @@ object KeepSource {
   // These may be old links
   val bulk = imports ++ Set(userCopied, unknown)
 
-  val discrete = Set(keeper, site, mobile, email, twitterSync)
+  // One-at-a-time keeps
+  val discrete = Set(keeper, site, mobile, email, twitterSync, slack)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala:318]: `slack` is not one-at-a-time

**Comment by andrewconner** [comment]
This fixes a bug where General libraries didn't get follower notifications. Fundamental problem was that side-effects were performed in 3 places, not always consistently. Now all post-keep side effects (analytics, data integrity, and user notification) are done together at the end.

File: kifi-backend/modules/common/app/com/keepit/model/Keep.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/Keep.scala:318]: Discussed this with Léo, we're starting with it here on purpose.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala [hunk 2]
```diff
@@ -80,7 +81,7 @@ class KeepInternerImpl @Inject() (
     log.info(s"[persistRawKeeps] persisting batch of ${rawKeeps.size} keeps")
     val newImportId = importId.getOrElse(UUID.randomUUID.toString)
 
-    val deduped = deDuplicateRawKeep(rawKeeps) map (_.copy(importId = Some(newImportId)))
+    val deduped = rawKeeps.distinctBy(_.url) map (_.copy(importId = Some(newImportId)))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala:84]: :+1: nice

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala [hunk 3]
```diff
@@ -150,38 +137,25 @@ class KeepInternerImpl @Inject() (
...
-          // It appears that when there's a problem, now, we're fetching from rover with ids that may not exist anymore.
-          // So, forcing the Try evaluation to make the db session fail, and then recatching it.
-          // Expect something more elegant soon. Signed, Andrew (May 22 2015)
-        }
-      }
-    }.toMap partition {
-      case (bm, res) => res.isSuccess
-    }
+
+    val (persisted, failed) = db.readWriteBatch(bms, attempts = 5) {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/KeepInterner.scala:150]: I bet you're asleep, but I'm going to try this again. It's about 5x faster, so lets us ingest keeps much faster. I think the previous issues were `onTransactionSuccess` leaking references that were later rolled back. It's removed now, and side effects (including rover asap fetches) are done at the end explicitly. Let me know if you think this is fine. @leogrim

**Pull request merged**